compiler/rustc_type_ir/src/macros.rs

```diff
@@ -1,54 +1,28 @@
-/// Used for types that are `Copy` and which **do not care arena
-/// allocated data** (i.e., don't need to be folded).
-macro_rules! TrivialTypeTraversalImpls {
-    ($($ty:ty,)+) => {
-        $(
-            impl<I: $crate::Interner> $crate::fold::TypeFoldable<I> for $ty {
-                fn try_fold_with<F: $crate::fold::FallibleTypeFolder<I>>(
-                    self,
-                    _: &mut F,
-                ) -> ::std::result::Result<Self, F::Error> {
-                    Ok(self)
-                }
+#[macro_export]
+macro_rules! noop_traversal_if_boring {
+    ($val:tt.try_fold_with($folder:expr)) => {{
+        use $crate::fold::SpecTypeFoldable as _;
+        $crate::noop_traversal_if_boring!($val.spec_try_fold_with($folder))
+    }};
+    ($val:tt.visit_with($visitor:expr)) => {{
+        use $crate::visit::SpecTypeVisitable as _;
+        $crate::noop_traversal_if_boring!($val.spec_visit_with($visitor))
+    }};
+    ($val:tt.$method:ident($traverser:expr)) => {{
+        let val = $val;
 
-                #[inline]
-                fn fold_with<F: $crate::fold::TypeFolder<I>>(
-                    self,
-                    _: &mut F,
-                ) -> Self {
-                    self
-                }
-            }
+        #[allow(unreachable_code)]
+        let p = 'p: {
+            break 'p ::core::marker::PhantomData;
 
-            impl<I: $crate::Interner> $crate::visit::TypeVisitable<I> for $ty {
-                #[inline]
-                fn visit_with<F: $crate::visit::TypeVisitor<I>>(
-                    &self,
-                    _: &mut F)
-                    -> ::std::ops::ControlFlow<F::BreakTy>
-                {
-                    ::std::ops::ControlFlow::Continue(())
-                }
+            #[inline(always)]
+            pub fn unreachable_phantom_constraint<T>(_: T) -> std::marker::PhantomData<T> {
+                unreachable!()
             }
-        )+
-    };
-}
 
-///////////////////////////////////////////////////////////////////////////
-// Atomic structs
-//
-// For things that don't carry any arena-allocated data (and are
-// copy...), just add them to this list.
+            unreachable_phantom_constraint(val)
+        };
```

Comment:
Have updated this locally to:

```rust
        fn phantom<T>(_: &T) -> ::core::marker::PhantomData<T> {
            ::core::marker::PhantomData
        }

        let val = $val;
        let p = phantom(&val);
```

but will hold of pushing again until there's some other reason to do so (there's bound to be something!).

Reply:
Actually, I recall now that we chose not to do this in order to guarantee that there's no function call in case it doesn't get optimised away.  I'll revert, but insert a `use ::core::marker::PhantomData` instead for readability.